pageserver/src/page_service.rs

```diff
@@ -899,165 +995,383 @@ impl PageServerHandler {
+        //
+
+        let PageServicePipeliningConfig {
+            max_batch_size,
+            protocol_pipelining_mode,
+        } = pipelining_config;
+
+        let cancel = self.cancel.clone();
+
+        //
+        // Create Reading future.
+        //
+
+        let (requests_tx, mut requests_rx) = tokio::sync::mpsc::channel(1);
+        let read_messages = {
+            let cancel = self.cancel.child_token();
+            let ctx = ctx.attached_child();
+            async move {
+                scopeguard::defer! {
+                    debug!("exiting");
+                }
+                let mut pgb_reader = pgb_reader;
+                loop {
+                    let msg = Self::pagestream_read_message(
+                        &mut pgb_reader,
+                        tenant_id,
+                        timeline_id,
+                        &mut timeline_handles,
+                        &cancel,
+                        &ctx,
+                        request_span.clone(),
+                    )
+                    .await?;
+                    let msg = match msg {
+                        Some(msg) => msg,
+                        None => {
+                            debug!("pagestream subprotocol end observed");
+                            break;
+                        }
+                    };
+                    match requests_tx.send(msg).await {
+                        Ok(()) => {}
+                        Err(tokio::sync::mpsc::error::SendError(_)) => {
+                            debug!("downstream is gone");
+                            break;
+                        }
                     }
-                };
+                }
+                Ok((pgb_reader, timeline_handles))
+            }
+        }
+        .instrument(tracing::info_span!("read_messages"));
+
+        //
+        // Create Batching future.
+        //
 
-                // Map handler result to protocol behavior.
-                // Some handler errors cause exit from pagestream protocol.
-                // Other handler errors are sent back as an error message and we stay in pagestream protocol.
-                for handler_result in handler_results {
-                    let response_msg = match handler_result {
-                        Err(e) => match &e {
-                            PageStreamError::Shutdown => {
-                                // If we fail to fulfil a request during shutdown, which may be _because_ of
-                                // shutdown, then do not send the error to the client.  Instead just drop the
-                                // connection.
-                                span.in_scope(|| info!("dropping connection due to shutdown"));
-                                return Err(QueryError::Shutdown);
+        enum BatchState {
+            Building(Option<Box<BatchedFeMessage>>),
+            ReadMessagesEnded(Option<Box<BatchedFeMessage>>),
```

Comment:
nit: would `PagestreamEnd` be a better name for this?

Reply:
Obsolete as of https://github.com/neondatabase/neon/pull/9851#discussion_r1859032178